File: samples/php/src/CoverageGate.php
<?php

declare(strict_types=1);

namespace DevKitSample;

use InvalidArgumentException;

/**
 * Evaluates per-file coverage against a project's bar. The sample mirrors the
 * kit's own coverage gate so its tests exercise real branches — a sample with
 * no branches would prove nothing about branch-level metrics.
 */
final class CoverageGate
{
    public function __construct(private readonly float $bar = 95.0)
    {
        if ($bar < 0.0 || $bar > 100.0) {
            throw new InvalidArgumentException('bar must be within 0..100');
        }
    }

    /** PASS, FAIL_LINES, or FAIL_BRANCHES — lines are reported first, like the kit's table. */
    public function verdict(float $lines, float $branches): string
    {
        foreach ([$lines, $branches] as $metric) {
            if ($metric < 0.0 || $metric > 100.0) {
                throw new InvalidArgumentException('coverage must be within 0..100');
            }
        }
        if ($lines < $this->bar) {
            return 'FAIL_LINES';
        }
        return $branches < $this->bar ? 'FAIL_BRANCHES' : 'PASS';
    }

    /** A touched file may not drop below its pre-change coverage, even above the bar. */
    public function regressed(float $before, float $after): bool
    {
        return $after < $before;
    }
}

File: samples/python/src/dev_kit_sample/coverage_gate.py
"""Evaluates per-file coverage against a project's bar.

The sample mirrors the kit's own coverage gate (and the PHP sample's class,
so the matrix rows stay comparable) — its tests exercise real branches, because
a branchless sample would prove nothing about branch-level metrics.
"""

from __future__ import annotations


class CoverageGate:
    """PASS / FAIL_LINES / FAIL_BRANCHES verdicts for one file's coverage."""

    def __init__(self, bar: float = 95.0) -> None:
        if not 0.0 <= bar <= 100.0:
            raise ValueError("bar must be within 0..100")
        self._bar = bar

    def verdict(self, lines: float, branches: float) -> str:
        """PASS, FAIL_LINES, or FAIL_BRANCHES — lines reported first, like the kit's table."""
        for metric in (lines, branches):
            if not 0.0 <= metric <= 100.0:
                raise ValueError("coverage must be within 0..100")
        if lines < self._bar:
            return "FAIL_LINES"
        return "FAIL_BRANCHES" if branches < self._bar else "PASS"

    def regressed(self, before: float, after: float) -> bool:
        """A touched file may not drop below its pre-change coverage, even above the bar."""
        return after < before
